File: battle_city/examples/nodejs/random.js
class Game {
    constructor(client) {
        this.client = client;
        this.firstTick = false;
        this.start = false;
    }

    receive(data) {
        switch (data.status) {
            case 'data':
                if (data.action === 'move') { return; } // to many data ;_;
                // do sth
            break;
            case 'game':
                switch (data.action) {
                    case 'start': this.start = true; break;
                    case 'over': this.start = false; break;
                }
            break; 
        }

        console.log(this._getColor(data), data, '\x1B[0m');
    }

    loop() {
        if (!this.firstTick) {
            this.send({action: 'greet', name: 'NODEJS'});
            this.firstTick = true;
        } 

        if (this.start) {
            let action = Math.floor(Math.random() * 3);
            switch (action) {
                case 0: // random speed
                    let speed = Math.round(Math.random() * 2);
                    this.send({action: 'set_speed', speed: speed});
                break;
                case 1: // random direction
                    let index = Math.floor(Math.random() * 4);
                    let direction = ['up', 'down', 'left', 'right'][index];
                    this.send({action: 'rotate', direction: direction});
                break;
                case 2: // SHOT SHOT SHOT
                    this.send({action: 'shoot'});
                break;
            }
        }

        this.callSoon(250);
    }

    _getColor(data) {
        switch (data.status) {
            case 'data':
                switch (data.action) {
                    case 'spawn': return '\x1B[92m'; break; // green color
                    case 'destroy': return '\x1B[93m'; break; // orange color
                    default: return '\x1B[0m'; // white color
                }
            break;
            case 'ERROR': return '\x1B[91m'; break; // red color
            case 'OK': return '\x1B[35m'; break; // purple color
            case 'game': return '\x1B[34m'; break; // blue color
            default: return '\x1B[0m'; // white color
        }
    }

    callSoon(time) {
        setTimeout(this.loop.bind(this), time);
    }

    send(data) {
        let message = JSON.stringify(data);
        this.client.write(message + '\n');
    }

}

let net = require('net');
let client = new net.Socket();
client.setNoDelay();
let game = new Game(client);

client.connect({port: 8888, host: '127.0.0.1'}, function() {
    console.log('\x1B[1mCONNECTED!\x1B[0m');
    game.loop();
});

let buffer = '';
client.on('data', function(raw) {
    let prev, next;
    let message = raw.toString('utf8');
    while ((next = message.indexOf('\n', prev)) > -1) {
      buffer += message.substring(prev, next);

      let data = JSON.parse(buffer);
      game.receive(data);

      buffer = '';
      prev = next + 1;
    }
    buffer += message.substring(prev);
});

client.on('close', function() {
	console.log('Connection closed');
});

File: battle_city/examples/ruby/random.rb
require 'socket'
require 'json'
require 'time'


class Game
    def initialize(socket)
        @start = false
        @first_tick = false
        @socket = socket
        @timestamp = 0.0
        @prng = Random.new
    end

    def receive(data)
        # ruby with no external libs doesnt have async code
        # so I must check timestamp and run loop
        
        new_timestamp = Time.now.to_f
        if new_timestamp - @timestamp > 0.25
            tick
            @timestamp = new_timestamp
        end

        case data['status'] 
        when 'data'
            if data['action'] == 'move' then return end  # too many data ;_;
        when 'game'
            case data['action']
            when 'start' then @start = true
            when 'over' then @start = false
            end
        end

        print _get_color(data), data
        puts "\e[0m"
    end

    def tick
        if !@first_tick
            send 'action' => 'greet', 'name' => 'RUBY'
            @first_tick = true
        end

        if @start 
            action = @prng.rand(0..2)
            case action
            when 0  # random speed
                speed = @prng.rand(0..2)
                send 'action' => 'set_speed', 'speed' => speed
            when 1  # random direction
                index = @prng.rand(0..3)
                direction = ['up', 'down', 'left', 'right'][index]
                send 'action' => 'rotate', 'direction' => direction
            when 2  # SHOT SHOT
                send 'action' => 'shoot'
            end
        end
    end

    def _get_color(data)
        case data['status'] 
        when 'data'
            case data['action']
            when 'spawn' then "\e[92m"  # green color
            when 'destroy' then "\e[93m"  # orange color
            else "\e[0m"  # white color
            end
        when 'ERROR' then "\e[91m"  # red color
        when 'OK' then "\e[35m"  # purple color
        when 'game' then "\e[34m"  # blue color
        else "\e[0m"  # white color
        end
    end

    def send(data)
        @socket.puts(data.to_json)
    end
end

socket = TCPSocket.open('127.0.0.1', 8888)
game = Game.new(socket)
game.tick

while line = socket.gets
    game.receive(JSON.parse!(line))
end
socket.close
